python/repositories.bzl

```diff
@@ -1,8 +1,164 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
 """This file contains macros to be called during WORKSPACE evaluation.
 
 For historic reasons, pip_repositories() is defined in //python:pip.bzl.
 """
 
+load("//python/private:toolchains_repo.bzl", "PLATFORMS", "toolchains_repo")
+load("//python/private:versions.bzl", "MINOR_MAPPING", "RELEASE_DATE", "RELEASE_URL", "TOOL_VERSIONS")
+
 def py_repositories():
     # buildifier: disable=print
     print("py_repositories is a no-op and is deprecated. You can remove this from your WORKSPACE file")
+
+########
+# Remaining content of the file is only used to support toolchains.
+########
+_DOC = "Fetch external tools needed for python toolchain."
+_ATTRS = {
+    "platform": attr.string(mandatory = True, values = PLATFORMS.keys()),
+    "python_version": attr.string(mandatory = True, values = TOOL_VERSIONS.keys() + MINOR_MAPPING.keys()),
+}
+
+def _python_repo_impl(rctx):
+    version = rctx.attr.python_version
+    if version in MINOR_MAPPING:
+        version = MINOR_MAPPING[version]
+    release_filename = "cpython-{version}-{platform}-pgo+lto-{release_date}.tar.zst".format(
+        release_date = RELEASE_DATE,
+        version = version,
+        platform = rctx.attr.platform,
+    )
+    url = "{release_url}/{release_filename}".format(
+        release_url = RELEASE_URL,
+        release_filename = release_filename,
+    )
+    rctx.download(
+        url = url,
+        integrity = TOOL_VERSIONS[version][rctx.attr.platform],
+        output = release_filename,
+    )
+    unzstd = rctx.which("unzstd")
+    if not unzstd:
+        url = rctx.attr._zstd_url.format(version = rctx.attr._zstd_version)
+        rctx.download_and_extract(
+            url = url,
+            sha256 = rctx.attr._zstd_sha256,
+        )
+        working_directory = "zstd-{version}".format(version = rctx.attr._zstd_version)
+        rctx.execute(
+            ["make", "--jobs=4"],
+            timeout = 600,
+            quiet = True,
+            working_directory = working_directory,
+        )
+        zstd = "{working_directory}/zstd".format(working_directory = working_directory)
+        unzstd = "./unzstd"
+        rctx.symlink(zstd, unzstd)
+
+    exec_result = rctx.execute([
+        "tar",
+        "--extract",
+        "--strip-components=2",
+        "--use-compress-program={unzstd}".format(unzstd = unzstd),
+        "--file={}".format(release_filename),
+    ])
+    if exec_result.return_code:
+        fail(exec_result.stderr)
+
+    build_content = """\
+# Generated by python/repositories.bzl
+
+load("@bazel_tools//tools/python:toolchain.bzl", "py_runtime_pair")
+
+filegroup(
+    name = "files",
+    srcs = glob(["bin/**", "lib/**", "include/**", "share/**"]),
```

Comment:
Be sure to exclude `__pycache__` and `*.pyc` files as that may introduce cache busts in production systems.

Reply:
Not sure I follow why the pre-compiled modules would introduce cache busts? Could you expand, please?

Reply:
In older versions of Python the generated bytecode was non-deterministic, but given we are fetching prebuilt versions this may be a non-issue as they wouldn't change, unless Python will continually write new `.pyc` files here?
We should add a test against the SHA of these files perhaps.

Reply:
Do you know what changed within those files? My guess is that the interpreter does indeed update these files again and they contain timestamps or absolute paths.

We should exclude `RECORD` files for the same reason then.

Reply:
I cannot find the logs right now, but my guess was that:
1. We download the python interpreter files and extract them.
2. We run pip_parse, which uses `bin/python` from the repository rule via the `python_interpreter_target` attribute.
3. This causes extra `.pyc` files being created during the `pip_parse`
4. Since the `.pyc` files are non-deterministic, we re-execute the Python test targets every time.

Having a minimal example should not be too difficult, but unfortunately I don't have time right now to help.

Reply:
@aignas could you try the latest state of this PR? I made the `lib` directory read-only, so it should be immutable. I wonder if you will still get the same error.